### ISSUE TITLE
File name completion changes default directory

### ISSUE THREAD

**Issue by maxchaos** [abo-abo/swiper#2610]
Hello. In the past few weeks, I noticed that when I execute a command (such as `org-insert-link` with a prefix arg) that uses `ivy` to simply complete and return the path to a file without visiting it, the value of `default-directory` in the current buffer changes to the parent directory of the selected file. I have verified that this is caused by `ivy-call` by temporarily commenting out line [1481](https://github.com/abo-abo/swiper/blob/db5a8940cf2a4324a1512c8f2a0e91475d1ac26f/ivy.el#L1481) of `ivy.el`. Besides that, I have no other clue as to what caused this problem.

**Comment by mcfarljm** [comment]
I am having trouble reproducing this.  Testing on db5a894, here is what I tried:

- Run `make plain` from the git root directory
- `M-x org-mode` to activate org-mode
- `C-u C-c C-l` (`org-insert-link` with prefix), and then select a file in a different directory
- Check the current directory using `C-h v default-directory` (or also just `C-x C-f` to see where it starts at)

When I try these steps, the value of `default-directory` does not change from the git root directory where I started emacs, as expected.  Can you confirm whether you still have the issue if you launch emacs without any additional packages (i.e., by running `make plain` from a copy of the swiper git repo)?

**Comment by maxchaos** [comment]
Indeed, when launching Emacs using `make plain` and following your instruction, the value of `default-directory` does not change. After playing a bit using `edebug`, I concluded that, when launching Emacs this way, the value that `action` assumes inside the call of `ivy-call` is `identity` and, as such, execution never reaches the branch which modifies the value of `default-directory`. When launching Emacs normally, it appears that `action` gets assigned a byte-compiled function which causes execution to enter the corresponding branch, resulting in this unwanted behavior.

Nevertheless, taking a closer look at that particular branch of `ivy-call`, modifying the value of `default-directory` using `setq` feels wrong because this way the new value persists beyond the dynamic scope of this function, and I think this function is not supposed to modify the environment of its caller. Correct me if I am wrong here, but it seems to me that the whole purpose of assigning a value to `default-directory` is to _temporarily_ modify the execution environment of the subsequent sexp, i.e., `(funcall action x)`. In that case, replacing
```
(setq default-directory (ivy-state-directory ivy-last))
(funcall action x)
```
with
```
(let ((default-directory
        (ivy-state-directory ivy-last)))
  (funcall action x))
```
seems more appropriate and resolves the issue at hand.

**Comment by basil-conto** [comment]
@maxchaos Doesn't the latter break <kbd>M-x</kbd>`cd`? See #2549.

**Comment by maxchaos** [comment]
You are right... In this case, I don't know what is the right way around this problem. I mean, is `ivy-call` supposed to alter the default directory when `action` is anything other than `identiy`?

**Comment by basil-conto** [comment]
Sorry, the logic of `ivy-call` is beyond me.

**Comment by mcfarljm** [comment]
I'm not sure why this would have to do with whether or not you launch emacs via `make plain`.  The value of `action` is being assigned based on the value of `ivy-last`, and this is changed dynamically as different ivy commands are run.  I think the problem is that `(ivy--get-action ivy-last)` should be producing `'identity` when you get into `ivy-call` via `C-u C-c C-l`.  When I test, that's what it's value is, and then you don't have the issue with changing the directory.  Maybe it has to do with the sequence of commands that you are using are something else about your setup?

**Comment by maxchaos** [comment]
I played with the debugger a bit more and managed to figure out what `(ivy--get-action ivy-last)` returns in my case. You see, I use [prescient](https://github.com/raxod502/prescient.el) for sorting completion candidates based on frequency and this package sets up things so `(ivy--get-action ivy-last)` returns some sort of byte-compiled wrapper around `identity` which first records the selection before returning it. I also bumped into a related issue on that repository (raxod502/prescient.el#61). So, I guess the question right now is whether `ivy-call` should blindly modify the value of `default-directory`? Is there a particular example of why this is necessary?
